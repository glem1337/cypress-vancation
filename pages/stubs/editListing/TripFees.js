export { default } from 'views/stubs/editListing/TripFees';
