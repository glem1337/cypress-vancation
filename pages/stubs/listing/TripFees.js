export { default } from 'views/stubs/listing/TripFees';
