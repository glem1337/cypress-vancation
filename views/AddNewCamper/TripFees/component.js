import { FormattedMessage } from 'react-intl';
import {
 Col, Row, Form, Divider, Skeleton,
} from 'antd';

import PropTypes from 'prop-types';

import ProgressBar from 'views/shared/ProgressBar';
import AddNewCamperBtnForm from 'views/shared/AddNewCamperBtnForm';

import GeneratorSection from './GeneratorSection';
import MileageSection from './MileageSection';
import CleaningsSection from './CleaningsSection';
import CustomFeesSection from './CustomFeesSection';
import EstimatedEarningCard from '../EstimatedEarningCard/container';

const TripFeesComponent = ({
  values,
  onMilesModeChange,
  handleSubmit,
  onBackButtonClick,
  onGeneratorExistenceChange,
  onGeneratorModeChange,
  addCustomFee,
  removeCustomFee,
  onFrequencyChanged,
  isLoading,
  isCamperExist,
  camperCompleteness,
}) => {
  const {
    generator,
    mileage,
    customFees,
    cleaningAndPreparationFee,
  } = values;

  if (!isCamperExist) {
    return (
      <div className="main-listing-container">
        <div className="mb-16 mb-md-24">
          <ProgressBar
            content={<FormattedMessage id="addNewCamper.ProgressBar" />}
            percent={camperCompleteness}
          />
        </div>
        <Row>
          <Col span={24}>
            <Skeleton active />
          </Col>
        </Row>
      </div>
    );
  }

  return (
    <>
      <div className="main-listing-container">
        <div className="mb-16 mb-md-24">
          <ProgressBar
            content={<FormattedMessage id="addNewCamper.ProgressBar" />}
            percent={camperCompleteness}
          />
        </div>
        <Form layout="vertical">
          <CleaningsSection cleaningAndPreparationFee={cleaningAndPreparationFee} />
          <Divider className="mt-24" />
          <Row>
            <Col span={24}>
              <MileageSection
                onMilesModeChange={onMilesModeChange}
                mode={mileage.mode}
              />
            </Col>
          </Row>
          <Divider className="mt-24" />
          <Row>
            <Col span={24}>
              <GeneratorSection
                hasGenerator={generator.hasGenerator}
                onGeneratorExistenceChange={onGeneratorExistenceChange}
                mode={generator.mode}
                onGeneratorModeChange={onGeneratorModeChange}
              />
            </Col>
          </Row>
          <Divider className="mt-24" />
          <CustomFeesSection
            addCustomFee={addCustomFee}
            customFees={Object.values(customFees)}
            removeCustomFee={removeCustomFee}
            onFrequencyChanged={onFrequencyChanged}
          />
        </Form>
        <EstimatedEarningCard isSlim />
      </div>
      <AddNewCamperBtnForm
        withBackBtn
        canSave={!isLoading}
        onSaveClick={handleSubmit}
        onBackClick={onBackButtonClick}
        isSubmitting={isLoading}
      />
    </>
  );
};

TripFeesComponent.propTypes = {
  values: PropTypes.shape().isRequired,
  onMilesModeChange: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onBackButtonClick: PropTypes.func.isRequired,
  onGeneratorExistenceChange: PropTypes.func.isRequired,
  onGeneratorModeChange: PropTypes.func.isRequired,
  addCustomFee: PropTypes.func.isRequired,
  removeCustomFee: PropTypes.func.isRequired,
  onFrequencyChanged: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isCamperExist: PropTypes.bool.isRequired,
  camperCompleteness: PropTypes.number.isRequired,
};

export default TripFeesComponent;
