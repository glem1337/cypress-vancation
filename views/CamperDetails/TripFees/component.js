import PropTypes from 'prop-types';
import { Row, Col } from 'antd';

import SkeletonText from 'views/shared/SkeletonText';
import Limits from './Limits';
import OwnerFees from './OwnerFees';

const TripFees = ({
  isCamperExist,
  isMileageLimited,
  availableMiles,
  overageMiles,
  hasGenerator,
  isGeneratorLimited,
  availableGeneratorHours,
  overageGenerator,
  cleaningFee,
  dumpingFee,
  fuelFee,
  lateDropOff,
  customFees,
  ownerFeesIsPresent,
}) => {
  if (!isCamperExist) {
    return (
      <Row gutter={24}>
        <Col md={12} className="mb-24 mb-md-0">
          <div className="skeleton__title w-100" />
          <SkeletonText rows={3} />
        </Col>
        <Col md={12}>
          <div className="skeleton__title w-100" />
          <SkeletonText rows={3} />
        </Col>
      </Row>
    );
  }

  return (
    <Row gutter={24}>
      <Limits
        isMileageLimited={isMileageLimited}
        availableMiles={availableMiles}
        overageMiles={overageMiles}
        hasGenerator={hasGenerator}
        isGeneratorLimited={isGeneratorLimited}
        availableGeneratorHours={availableGeneratorHours}
        overageGenerator={overageGenerator}
      />
      {ownerFeesIsPresent && (
        <OwnerFees
          cleaningFee={cleaningFee}
          customFees={customFees}
          dumpingFee={dumpingFee}
          fuelFee={fuelFee}
          lateDropOff={lateDropOff}
        />
      )}
    </Row>
  );
};

TripFees.propTypes = {
  isCamperExist: PropTypes.bool.isRequired,
  isMileageLimited: PropTypes.bool.isRequired,
  availableMiles: PropTypes.number.isRequired,
  overageMiles: PropTypes.number.isRequired,
  hasGenerator: PropTypes.bool.isRequired,
  isGeneratorLimited: PropTypes.bool.isRequired,
  availableGeneratorHours: PropTypes.number.isRequired,
  overageGenerator: PropTypes.number.isRequired,
  cleaningFee: PropTypes.number.isRequired,
  dumpingFee: PropTypes.number.isRequired,
  fuelFee: PropTypes.number.isRequired,
  lateDropOff: PropTypes.number.isRequired,
  customFees: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  ownerFeesIsPresent: PropTypes.bool.isRequired,
};

export default TripFees;
