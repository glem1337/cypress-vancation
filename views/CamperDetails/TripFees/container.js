import React from 'react';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import { connect } from 'react-redux';

import { DEFAULT_VALUES } from 'constants/camperTripFees';

import isPresent from 'utils/isPresent';

import {
  camperSelector,
  isCamperExistSelector,
} from 'state/concepts/camper/selectors';

import TripFeesComponent from './component';

class TripFees extends React.Component {
  static defaultProps = {
    camper: null,
  };

  static propTypes = {
    camper: PropTypes.shape(),
  };

  get isMileageLimited() {
    const { camper } = this.props;

    return R.pathOr(false, ['tripFee', 'tripFeeMileage', 'limit'], camper);
  }

  get availableMiles() {
    const { camper } = this.props;

    return R.pathOr(
      DEFAULT_VALUES.MILEAGE.INCLUDED,
      ['tripFee', 'tripFeeMileage', 'available'],
      camper,
    );
  }

  get overageMiles() {
    const { camper } = this.props;

    return R.pathOr(
      DEFAULT_VALUES.MILEAGE.OVERAGE,
      ['tripFee', 'tripFeeMileage', 'overage'],
      camper,
    );
  }

  get hasGenerator() {
    const { camper } = this.props;

    return isPresent(camper?.tripFee?.tripFeeGenerator);
  }

  get isGeneratorLimited() {
    const { camper } = this.props;

    return R.pathOr(false, ['tripFee', 'tripFeeGenerator', 'limit'], camper);
  }

  get availableGeneratorHours() {
    const { camper } = this.props;

    return R.pathOr(
      DEFAULT_VALUES.GENERATOR.INCLUDED,
      ['tripFee', 'tripFeeGenerator', 'available'],
      camper,
    );
  }

  get overageGenerator() {
    const { camper } = this.props;

    return R.pathOr(
      DEFAULT_VALUES.GENERATOR.OVERAGE,
      ['tripFee', 'tripFeeGenerator', 'overage'],
      camper,
    );
  }

  get cleaningFee() {
    const { camper } = this.props;

    return R.pathOr(0, ['tripFee', 'cleaning'], camper);
  }

  get dumpingFee() {
    const { camper } = this.props;

    return R.pathOr(0, ['tripFee', 'dumpingFee'], camper);
  }

  get fuelFee() {
    const { camper } = this.props;

    return R.pathOr(0, ['tripFee', 'flue'], camper);
  }

  get lateDropOff() {
    const { camper } = this.props;

    return R.pathOr(0, ['tripFee', 'lateDropOff'], camper);
  }

  get customFees() {
    const { camper } = this.props;

    return R.pathOr([], ['tripFee', 'customFees'], camper);
  }

  get ownerFeesIsPresent() {
    const {
      cleaningFee,
      dumpingFee,
      fuelFee,
      lateDropOff,
      customFees,
    } = this;

    return Boolean(
      cleaningFee
        || dumpingFee
        || fuelFee
        || lateDropOff
        || isPresent(customFees),
    );
  }

  render() {
    return (
      <TripFeesComponent
        {...this.props}
        isMileageLimited={this.isMileageLimited}
        availableMiles={this.availableMiles}
        overageMiles={this.overageMiles}
        hasGenerator={this.hasGenerator}
        isGeneratorLimited={this.isGeneratorLimited}
        availableGeneratorHours={this.availableGeneratorHours}
        overageGenerator={this.overageGenerator}
        cleaningFee={this.cleaningFee}
        dumpingFee={this.dumpingFee}
        fuelFee={this.fuelFee}
        lateDropOff={this.lateDropOff}
        customFees={this.customFees}
        ownerFeesIsPresent={this.ownerFeesIsPresent}
      />
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  camper: camperSelector(state, ownProps?.camperId),
  isCamperExist: isCamperExistSelector(state, ownProps?.camperId),
});

export { TripFees as TripFeesContainer };
export default connect(mapStateToProps)(TripFees);
