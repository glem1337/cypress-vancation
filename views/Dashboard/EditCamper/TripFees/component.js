import PropTypes from 'prop-types';
import {
 Col, Row, Form, Divider, Skeleton,
} from 'antd';

import EditCamperLayout from 'views/layouts/EditCamper';

import GeneratorSection from 'views/AddNewCamper/TripFees/GeneratorSection';
import MileageSection from 'views/AddNewCamper/TripFees/MileageSection';
import CleaningsSection from 'views/AddNewCamper/TripFees/CleaningsSection';
import CustomFeesSection from 'views/AddNewCamper/TripFees/CustomFeesSection';
import DumpingSection from './DumpingSection';

const TripFeesComponent = ({
  values,
  onMilesModeChange,
  handleSubmit,
  onGeneratorExistenceChange,
  onGeneratorModeChange,
  addCustomFee,
  removeCustomFee,
  onFrequencyChanged,
  isLoading,
  isCamperExist,
  leavePagePrepare,
  intl: { formatMessage },
}) => {
  const { generator, mileage, customFees, cleaningAndPreparationFee } = values;

  return (
    <EditCamperLayout
      hasFooter
      canSave={!isLoading}
      onSave={handleSubmit}
      isLoading={isLoading}
      leavePageMethod={leavePagePrepare}
    >
      <div className="container">
        {isCamperExist ? (
          <Form layout="vertical">
            <Row>
              <Col span={24}>
                <CleaningsSection
                  cleaningAndPreparationFee={cleaningAndPreparationFee}
                />
              </Col>
              <Divider className="mt-24" />
              <Col span={24}>
                <MileageSection
                  onMilesModeChange={onMilesModeChange}
                  mode={mileage.mode}
                />
              </Col>
              <Divider className="mt-24" />
              <Col span={24}>
                <GeneratorSection
                  hasGenerator={generator.hasGenerator}
                  onGeneratorExistenceChange={onGeneratorExistenceChange}
                  mode={generator.mode}
                  onGeneratorModeChange={onGeneratorModeChange}
                />
              </Col>
              <Divider className="mt-24" />
              <Col span={24}>
                <DumpingSection formatMessage={formatMessage} />
              </Col>
              <Divider className="mt-24" />
              <Col span={24}>
                <CustomFeesSection
                  addCustomFee={addCustomFee}
                  customFees={Object.values(customFees)}
                  removeCustomFee={removeCustomFee}
                  onFrequencyChanged={onFrequencyChanged}
                />
              </Col>
            </Row>
          </Form>
        ) : (
          <Skeleton active />
        )}
      </div>
    </EditCamperLayout>
  );
};

TripFeesComponent.propTypes = {
  values: PropTypes.shape().isRequired,
  onMilesModeChange: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onGeneratorExistenceChange: PropTypes.func.isRequired,
  onGeneratorModeChange: PropTypes.func.isRequired,
  addCustomFee: PropTypes.func.isRequired,
  removeCustomFee: PropTypes.func.isRequired,
  onFrequencyChanged: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isCamperExist: PropTypes.bool.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
  leavePagePrepare: PropTypes.func.isRequired,
};

export default TripFeesComponent;
