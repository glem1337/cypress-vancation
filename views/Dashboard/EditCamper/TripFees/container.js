import React from 'react';
import { withFormik } from 'formik';
import { compose } from 'ramda';
import { withRouter } from 'next/router';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { v4 as uuid } from 'uuid';
import * as R from 'ramda';
import { injectIntl } from 'react-intl';

import { CAMPER_INCLUSION, TRIP_FEES } from 'constants/camper';
import { DEFAULT_VALUES } from 'constants/camperTripFees';

import { leavePage, leavePagePrepare } from 'utils/camper/leavePageHelper';
import { handleSubmitWithProps } from 'utils/form/handleSubmit';
import {
  deleteCustomTripFee as deleteCustomTripFeeAction,
  fetchCamper,
  updateTripFees,
} from 'state/concepts/camper/actions';
import {
  camperSelector,
  isCamperExistSelector,
} from 'state/concepts/camper/selectors';
import { updateTripFeesEndpoint } from 'state/concepts/camper/endpoints';
import {
  showModal as showModalAction,
  hideModal as hideModalAction,
} from 'state/modal/actions';
import { loadingSelector } from 'state/data/selectors';
import { dataDeleteEntity } from 'state/data/actions';

import yup from 'lib/yupLocalised';

import TripFeesComponent from './component';

class TripFees extends React.Component {
  static propTypes = {
    values: PropTypes.shape().isRequired,
    errors: PropTypes.shape().isRequired,
    touched: PropTypes.shape().isRequired,
    setFormikState: PropTypes.func.isRequired,
    setFieldValue: PropTypes.func.isRequired,
    router: PropTypes.shape().isRequired,
    camperId: PropTypes.string.isRequired,
    deleteCustomTripFee: PropTypes.func.isRequired,
    showModal: PropTypes.func.isRequired,
    hideModal: PropTypes.func.isRequired,
    intl: PropTypes.shape().isRequired,
    isLoading: PropTypes.bool,
    isValid: PropTypes.bool.isRequired,
    dirty: PropTypes.bool.isRequired,
    handleSubmit: PropTypes.func.isRequired,
    validateForm: PropTypes.func.isRequired,
  };

  static defaultProps = {
    isLoading: false,
  };

  leavePage = leavePage.bind(this);

  leavePagePrepare = leavePagePrepare.bind(this);

  static getInitialProps = async (ctx) => {
    const camperId = ctx.query.camper;

    ctx.store.dispatch(
      dataDeleteEntity({
        kind: 'tripFee',
      }),
    );

    const inclusions = [
      CAMPER_INCLUSION.TRIP_FEE.INDEX,
      CAMPER_INCLUSION.TRIP_FEE.TRIP_FEE_MILEAGE,
      CAMPER_INCLUSION.TRIP_FEE.TRIP_FEE_GENERATOR,
      CAMPER_INCLUSION.TRIP_FEE.CUSTOM_FEES,
      // Have to add (this) due to backend has changed architecture of pricing info
      CAMPER_INCLUSION.PRICING_INFO.WEEK_NIGHT_PRICE,
      CAMPER_INCLUSION.PRICING_INFO.INDEX,
    ];

    ctx.store.dispatch(fetchCamper(camperId, inclusions.join(',')));

    return { camperId };
  };

  static isUnlimitedMileageMode = (mode) => mode === TRIP_FEES.MILEAGE_VALUES.UNLIMITED;

  static isUnlimitedGeneratorMode = (mode) => mode === TRIP_FEES.GENERATOR_VALUES.UNLIMITED;

  static determineInitialValues = (props) => {
    const tripFee = R.pathOr({}, ['camper', 'tripFee'], props);

    // Cleaning
    const cleaningAndPreparationFee = tripFee.cleaning || '';

    // Mileage
    const mileageIncluded = R.path(['tripFeeMileage', 'available'], tripFee);
    const mileageOverage = R.path(['tripFeeMileage', 'overage'], tripFee);
    const mileageLimit = R.path(['tripFeeMileage', 'limit'], tripFee);
    const mileage = {
      mode: mileageLimit
        ? TRIP_FEES.MILEAGE_VALUES.LIMITED
        : TRIP_FEES.MILEAGE_VALUES.UNLIMITED,
      included: mileageIncluded || DEFAULT_VALUES.MILEAGE.INCLUDED,
      overage: mileageOverage || DEFAULT_VALUES.MILEAGE.OVERAGE,
    };

    // Generator
    const hasGenerator = R.compose(
      R.not,
      R.isNil(),
      R.path(['tripFeeGenerator', 'presence']),
    )(tripFee);
    const generatorLimit = R.path(['tripFeeGenerator', 'limit'], tripFee);
    const generatorIncluded = R.path(
      ['tripFeeGenerator', 'available'],
      tripFee,
    );
    const generatorOverage = R.path(['tripFeeGenerator', 'overage'], tripFee);
    const generator = {
      hasGenerator,
      mode:
        generatorLimit === true
          ? TRIP_FEES.GENERATOR_VALUES.LIMITED
          : TRIP_FEES.GENERATOR_VALUES.UNLIMITED,
      included: generatorIncluded || DEFAULT_VALUES.GENERATOR.INCLUDED,
      overage: generatorOverage || DEFAULT_VALUES.GENERATOR.OVERAGE,
    };

    // Custom fees
    const customFees = R.defaultTo([], tripFee.customFees).reduce(
      (acc, item) => R.assoc(
          item.id,
          {
            id: item.id,
            name: item.name,
            price: item.price,
            frequency: item.frequency,
            fromServer: true,
          },
          acc,
        ),
      {},
    );

    return {
      cleaningAndPreparationFee,
      mileage,
      generator,
      customFees,
      dumpingFee: tripFee?.dumpingFee || TRIP_FEES.DUMPING_VALUES.DUMPING,
      flue: tripFee?.flue || TRIP_FEES.DUMPING_VALUES.FLUE,
      lateDropOff: tripFee?.lateDropOff || TRIP_FEES.DUMPING_VALUES.LATE_DROP_OFF,
    };
  };

  static getCustomFeeSchema = (obj) => {
    const schema = {};

    const keys = Object.keys(obj);

    for (let i = 0; i < keys.length; i += 1) {
      const key = keys[i];

      schema[key] = yup.object({
        name: yup.string().required(),
        price: yup
          .number()
          .moreThan(0, { id: 'validations.cantBeValue', values: { value: 0 } })
          .required(),
        frequency: yup.string(),
      });
    }

    return yup.object(schema);
  };

  componentDidUpdate(prevProps) {
    const {
      values: { customFees },
    } = this.props;

    // Scroll to the bottom
    if (
      Object.keys(customFees).length
      !== Object.keys(prevProps.values.customFees).length
    ) {
      this.scrollContainerToBottom();
    }
  }

  scrollContainerToBottom = () => {
    const scrollingElement = document.querySelector('.edit-list-wrap');
    scrollingElement.scrollTop = scrollingElement.scrollHeight;

    window.scrollTo(0, scrollingElement.scrollHeight);

    return true;
  };

  onMilesModeChange = (event) => {
    const { values, errors, touched, setFormikState } = this.props;

    const { value } = event.target;

    const formikState = {
      errors: {
        ...errors,
        mileage: errors.mileage || {},
      },
      touched: {
        ...touched,
        mileage: touched.mileage || {},
      },
      values: {
        ...values,
        mileage: {
          ...values.mileage,
          mode: value,
        },
      },
    };

    if (value === TRIP_FEES.MILEAGE_VALUES.UNLIMITED) {
      formikState.values.mileage.included = DEFAULT_VALUES.MILEAGE.INCLUDED;
      formikState.values.mileage.overage = DEFAULT_VALUES.MILEAGE.OVERAGE;

      delete formikState.errors.mileage;
      delete formikState.touched.mileage;
    }

    setFormikState(formikState);

    return formikState;
  };

  onGeneratorExistenceChange = (isChecked) => {
    const { values, errors, touched, setFormikState } = this.props;

    const formikState = {
      errors: {
        ...errors,
        generator: errors.generator || {},
      },
      touched: {
        ...touched,
        generator: touched.generator || {},
      },
      values: {
        ...values,
        generator: {
          ...values.generator,
          hasGenerator: isChecked,
        },
      },
    };

    if (!isChecked) {
      formikState.values.generator.mode = TRIP_FEES.GENERATOR_VALUES.UNLIMITED;
      formikState.values.generator.included = DEFAULT_VALUES.GENERATOR.INCLUDED;
      formikState.values.generator.overage = DEFAULT_VALUES.GENERATOR.OVERAGE;

      delete formikState.errors.generator;
      delete formikState.touched.generator;
    }

    setFormikState(formikState);

    return formikState;
  };

  onGeneratorModeChange = (event) => {
    const { values, errors, touched, setFormikState } = this.props;

    const { value } = event.target;

    const formikState = {
      errors: {
        ...errors,
        generator: errors.generator || {},
      },
      touched: {
        ...touched,
        generator: touched.generator || {},
      },
      values: {
        ...values,
        generator: {
          ...values.generator,
          mode: value,
        },
      },
    };

    if (value === TRIP_FEES.GENERATOR_VALUES.UNLIMITED) {
      formikState.values.generator.included = DEFAULT_VALUES.GENERATOR.INCLUDED;
      formikState.values.generator.overage = DEFAULT_VALUES.GENERATOR.OVERAGE;

      delete formikState.errors.generator;
      delete formikState.touched.generator;
    }

    setFormikState(formikState);

    return formikState;
  };

  addCustomFee = () => {
    const { setFieldValue, values } = this.props;

    const id = uuid();

    const customFees = {
      ...values.customFees,
      [id]: {
        id,
        name: '',
        price: '',
        frequency: TRIP_FEES.FREQUENCY_OPTIONS.PER_DAY.VALUE,
      },
    };

    setFieldValue('customFees', customFees);
  };

  removeCustomFeePrepare = (customFee) => () => {
    const {
      showModal,
      hideModal,
      intl: { formatMessage },
    } = this.props;

    showModal({
      modalType: 'REMOVE_SIGNATURE_MODAL',
      modalProps: {
        id: customFee.id,
        title: formatMessage({ id: 'addNewCamper.tripFees.removeModal.title' }),
        subTitle: formatMessage({
          id: 'addNewCamper.tripFees.removeModal.subTitle',
        }),
        cancel: hideModal,
        remove: this.removeCustomFee(customFee),
      },
    });
  };

  removeCustomFee = (customFee) => () => {
    const {
 setFieldValue, values, deleteCustomTripFee, camperId, hideModal,
} = this.props;

    if (customFee.fromServer) {
      deleteCustomTripFee({
        feeId: customFee.id,
        camperId,
      });
    }

    const customFees = R.omit([customFee.id], values.customFees);

    setFieldValue('customFees', customFees);
    if (!customFee.fromServer) {
      hideModal();
    }
  };

  onFrequencyChanged = (feeId) => (value) => {
    const { setFieldValue, values } = this.props;

    const newCustomFeeSection = R.assocPath(
      [feeId, 'frequency'],
      value,
      values.customFees,
    );

    setFieldValue('customFees', newCustomFeeSection);
  };

  render() {
    return (
      <TripFeesComponent
        {...this.props}
        onMilesModeChange={this.onMilesModeChange}
        onGeneratorModeChange={this.onGeneratorModeChange}
        onGeneratorExistenceChange={this.onGeneratorExistenceChange}
        addCustomFee={this.addCustomFee}
        removeCustomFee={this.removeCustomFeePrepare}
        onFrequencyChanged={this.onFrequencyChanged}
        leavePagePrepare={this.leavePagePrepare}
      />
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  isCamperExist: isCamperExistSelector(state, ownProps?.camperId),
  camper: camperSelector(state, ownProps?.camperId),
  isLoading: loadingSelector(state, updateTripFeesEndpoint.endpoint),
});

const mapDispatchToProps = {
  onSubmit: updateTripFees,
  deleteCustomTripFee: deleteCustomTripFeeAction,
  showModal: showModalAction,
  hideModal: hideModalAction,
};

export { TripFees as TripFeesContainer };
export default compose(
  withRouter,
  injectIntl,
  connect(mapStateToProps, mapDispatchToProps),
  withFormik({
    enableReinitialize: true,
    mapPropsToValues: TripFees.determineInitialValues,
    validationSchema: yup.object().shape({
      cleaningAndPreparationFee: yup.number().moreThan(-1, {
        id: 'validations.cantBeValue',
        values: { value: 'negative' },
      }),
      mileage: yup.object().shape({
        included: yup.number().when('mode', {
          is: TripFees.isUnlimitedMileageMode,
          then: yup.number(),
          otherwise: yup
            .number()
            .moreThan(0, {
              id: 'validations.cantBeValue',
              values: { value: 0 },
            })
            .required(),
        }),
        overage: yup.number().when('mode', {
          is: TripFees.isUnlimitedMileageMode,
          then: yup.number(),
          otherwise: yup
            .number()
            .moreThan(0, {
              id: 'validations.cantBeValue',
              values: { value: 0 },
            })
            .required(),
        }),
      }),
      generator: yup.object().shape({
        included: yup.number().when('mode', {
          is: TripFees.isUnlimitedGeneratorMode,
          then: yup.number(),
          otherwise: yup
            .number()
            .moreThan(0, {
              id: 'validations.cantBeValue',
              values: { value: 0 },
            })
            .required(),
        }),
        overage: yup.number().when('mode', {
          is: TripFees.isUnlimitedGeneratorMode,
          then: yup.number(),
          otherwise: yup
            .number()
            .moreThan(0, {
              id: 'validations.cantBeValue',
              values: { value: 0 },
            })
            .required(),
        }),
      }),
      customFees: yup.lazy(TripFees.getCustomFeeSchema),
    }),
    handleSubmit: handleSubmitWithProps(['camperId']),
  }),
)(TripFees);
