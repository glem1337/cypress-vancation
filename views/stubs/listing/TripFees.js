import {
  Alert,
  Button,
  Col,
  Input,
  Radio,
  Row,
  Switch,
  Tag,
  Divider,
  Form,
} from 'antd';
import CustomFeesBlock from './components/CustomFeesBlock';
import LayoutListing from './LayoutListing';

const TripFees = () => (
  <LayoutListing>
    <Form layout="vertical">
      <Row>
        <Col lg={16}>
          <h1 className="text-headline mb-8">
            Trip fees
          </h1>
          <p className="mb-24">
            Set mileage, generator, and cleaning / prep fees.
          </p>
          <div className="mb-8 text-subheader font-700">
            Cleaning and preparation fee
          </div>
          <p className="mb-16">
            Set the amount charged to your renter for camper cleaning and preparation.
            Typically this fee includes standard cleaning, fuel and water fill up, grey
            water draining, battery storage is charged, etc.
          </p>
          <Row gutter={24}>
            <Col md={12}>
              <Form.Item
                label={<span className="main-input__label">Amount</span>}
              >
                <Input id="Amount" placeholder="$0" />
              </Form.Item>
            </Col>
            <Col md={21}>
              <Alert
                type="warning"
                className="mb-24"
                showIcon
                icon={<i className="icon icon-alert" />}
                message="Vancation recommends adding a cleaning and preparation fee of $50-100."
              />
            </Col>
          </Row>
        </Col>
      </Row>
      <Divider className="mt-0" />
      <Row>
        <Col span={24}>
          <Row>
            <Col lg={16}>
              <div className="mb-8 text-subheader font-700">
                Mileage
              </div>
              <p className="mb-24">
                How many miles can your renter drive per day?
              </p>
              <Form.Item>
                <Radio.Group defaultValue="a" size="large">
                  <div className="mb-20">
                    <Radio value="Free">
                      <span className="ml-12">
                        Allow unlimited miles
                      </span>
                    </Radio>
                  </div>
                  <div className="mb-20">
                    <Radio value="rates">
                      <span className="ml-12">
                        Limit the number of miles
                      </span>
                    </Radio>
                  </div>
                </Radio.Group>
              </Form.Item>
              <Row gutter={24}>
                <Col md={12}>
                  <div className="main-input-wrap-addtxt">
                    <Form.Item
                      label={<span className="main-input__label">Included miles</span>}
                    >
                      <Input id="Included" placeholder="0" />
                      <span className="main-input__add-txt">
                        per day
                      </span>
                    </Form.Item>
                  </div>
                </Col>
                <Col md={12}>
                  <div className="main-input-wrap-addtxt">
                    <Form.Item
                      label={<span className="main-input__label">Overage charge</span>}
                    >
                      <Input id="Overage" placeholder="$0" />
                      <span className="main-input__add-txt">
                        per mile
                      </span>
                    </Form.Item>
                  </div>
                </Col>
              </Row>
            </Col>
          </Row>
        </Col>
      </Row>
      <Divider />
      <Row>
        <Col span={24}>
          <Row>
            <Col lg={16}>
              <div className="mb-8 text-subheader font-700">
                Generator rates
              </div>
              <p className="mb-24">
                If your vehicle has a generator please set your hourly usage rates here.
              </p>
              <div className="d-flex align-items-center mb-16">
                <Form.Item>
                  <Switch />
                  <span className="ml-8">My camper has a generator</span>
                </Form.Item>
              </div>
              <Form.Item>
                <Radio.Group defaultValue="a" size="large">
                  <div className="mb-20">
                    <Radio value="Free">
                      <span className="ml-12">Allow unlimited generator hours</span>
                      <Tag
                        className="mt-12 mt-md-0 ml-24 ml-md-8"
                        color="success"
                        icon={<i className="icon icon-like-f" />}
                      >
                        Vancation recommended
                      </Tag>
                    </Radio>
                  </div>
                  <div className="mb-20">
                    <Radio value="rates">
                      <span className="ml-12">Set my own hourly rate</span>
                    </Radio>
                  </div>
                </Radio.Group>
              </Form.Item>
              <Row gutter={24}>
                <Col md={12}>
                  <div className="main-input-wrap-addtxt">
                    <Form.Item
                      label={<span className="main-input__label">Included hours</span>}
                    >
                      <Input id="hours" placeholder="0" />
                      <span className="main-input__add-txt">per day</span>
                    </Form.Item>
                  </div>
                </Col>
                <Col md={12}>
                  <div className="main-input-wrap-addtxt">
                    <Form.Item
                      label={<span className="main-input__label">Overage charge</span>}
                    >
                      <Input id="charge" placeholder="0" />
                      <span className="main-input__add-txt">per hours</span>
                    </Form.Item>
                  </div>
                </Col>
              </Row>
            </Col>
          </Row>
        </Col>
      </Row>
      <Divider />
      <Row>
        <Col lg={16}>
          <div className="mb-8 text-subheader font-400">
            Custom fees
          </div>
          <p className="mb-24">
            Add custom fees if you want.
          </p>
          <CustomFeesBlock />
          <CustomFeesBlock />
          <Button
            className="mb-24"
            type="secondary"
          >
            Add custom fee
          </Button>
        </Col>
      </Row>
    </Form>
  </LayoutListing>
);

export default TripFees;
