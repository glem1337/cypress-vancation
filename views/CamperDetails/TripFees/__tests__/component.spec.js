import { shallow } from 'enzyme';

import TripFees from '../component';

describe('TripFees component tests', () => {
  const props = {
    isCamperExist: true,
    isMileageLimited: true,
    availableMiles: 1,
    overageMiles: 2,
    hasGenerator: true,
    isGeneratorLimited: true,
    availableGeneratorHours: 3,
    overageGenerator: 4,
    cleaningFee: 5,
    dumpingFee: 6,
    fuelFee: 7,
    lateDropOff: 8,
    customFees: [
      {
        id: '1',
      },
    ],
    ownerFeesIsPresent: true,
  };

  const component = shallow(<TripFees {...props} />);

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when camper doesn`t exist', () => {
    component.setProps({
      isCamperExist: false,
    });

    expect(component).toMatchSnapshot();
  });
});
