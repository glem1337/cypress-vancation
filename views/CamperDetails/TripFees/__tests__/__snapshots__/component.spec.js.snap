// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripFees component tests matches snapshot 1`] = `
<Row
  gutter={24}
>
  <Limits
    availableGeneratorHours={3}
    availableMiles={1}
    hasGenerator={true}
    isGeneratorLimited={true}
    isMileageLimited={true}
    overageGenerator={4}
    overageMiles={2}
  />
  <OwnerFees
    cleaningFee={5}
    customFees={
      Array [
        Object {
          "id": "1",
        },
      ]
    }
    dumpingFee={6}
    fuelFee={7}
    lateDropOff={8}
  />
</Row>
`;

exports[`TripFees component tests matches snapshot when camper doesn\`t exist 1`] = `
<Row
  gutter={24}
>
  <Col
    className="mb-24 mb-md-0"
    md={12}
  >
    <div
      className="skeleton__title w-100"
    />
    <SkeletonText
      rows={3}
    />
  </Col>
  <Col
    md={12}
  >
    <div
      className="skeleton__title w-100"
    />
    <SkeletonText
      rows={3}
    />
  </Col>
</Row>
`;
